### ISSUE TITLE
Lux correction formula application start (100 or 1000 lx)

### ISSUE THREAD

**Issue by bjorja** [eldruin/veml6030-rs#2]
`convert_raw_als_to_lux` only applies the low gain correction formula from >1000 lx, however according to page 10 (20-sept-2019 design application document), see figure 10, it is applicable from >100 lx.

Or was this implemented as-is because page 10 sates:  

> The VEML7700 shows good linear behaviour for lux levels from 0.007 lx to about 1 klx. Illumination values higher than 1000 lx show non-linearity.

However if you zoom-in, quite a bit, on fig 9 once can start to see (minor) deviations from before 1000 lx.

**Comment by eldruin** [comment]
I only apply it to values > 1000lx because the manual states so, as you quoted:

> The VEML7700 shows good linear behaviour for lux levels from 0.007 lx to about 1 klx. Illumination values higher than 1000 lx show non-linearity.

According to Figure 10, the correction formula seems to yield good results for > 100lx as well, though. 
I suppose this value is a compromise, since the correction also has a computation/energy [cost](https://github.com/eldruin/veml6030-rs/blob/30d5aa78ef8ea443c8acde7e1d021bda6ce4f6e6/src/correction.rs#L45) as well as precision requirements/numerical stability risks.

Have you run into measurement deviations?

**Comment by bjorja** [comment]
Did not yet properly test how (if) significant the difference is, i was just checking the design application document with the driver implementation and noticed the inconsistency.

**Comment by bjorja** [comment]
Did a quick calculation (no need actually measure), with the maximum error defined as: (Lux_VEML - CORRECTED_Lux) / Lux_VEML = -7.5% for 1000 lx and -1.0% for 100 lx (rounded). It very gradually moves from 1.0 -> 7.5% for 100 -> 1000 lx. (applies to gain 1/4, 1/8 across all integration times)

Provided i did not make any silly mistakes.

The formula is just a polynomic trend line with excel (generated from data in fig 9), they can start from wherever they like. There might indeed be energy concerns depending on the application (possibly indeed also numerical stability); and they do provide other correction formula (same polynomic trend line but in a different range) if so desired (e.g. start from 1000 lx instead of 100 lx if one is fine with the 1-7.5% accuracy hit).

**Comment by eldruin** [comment]
Thanks for running the math on it.
What I could do is to change the code a bit so that the `convert_raw_als_to_lux` function always does the correction and the `read_lux` is the one that decides to apply it or not depending on the value being > 1000 lux.
Then, if the users want to apply the correction starting in another range, they can do `read_raw()` and then call `convert_raw_als_to_lux`. Then they can decide on an accuracy deviation limit. Something like: we want to apply the correction starting from a theoretical deviation of 2% and skip it otherwise.

**Comment by bjorja** [comment]
Makes sense. If so i would presume the client becomes responsible for the gain check as well when not using `read_lux` (or will you automatically return the input if != 1/4 or 1/8).

I think the flow should become: `read_raw()`, `convert_to_lux(als)`  (== Lux_VEML), if desired `apply_lux_correction(lx)` (now internally called `correct_high_lux`); where `read_lux` does the same internally with gain and cut off check. You can also provide an `convert_to_lux` version taking gain and IT as well (as you do now for `convert_raw_als_to_lux` as well).

(to be honest if i would use a cut off different from 100 lx i would contact Vishey for a different correction formula, hoping it is cheaper to compute, and apply it myself client side but for that one would need a `convert_to_lux(als)` function exposed (or duplicate the logic client side))
